llm_studio/src/utils/modeling_utils.py

```diff
@@ -214,59 +301,66 @@ def get_optimizer(model: torch.nn.Module, cfg: Any) -> torch.optim.Optimizer:
     Returns:
         Optimizer
     """
-
-    no_decay = ["bias", "LayerNorm.weight"]
-    differential_layers = cfg.training.differential_learning_rate_layers
-    optimizer = Optimizers.get(cfg.training.optimizer)(
-        [
-            {
-                "params": [
-                    param
-                    for name, param in model.named_parameters()
-                    if (not any(layer in name for layer in differential_layers))
-                    and (not any(nd in name for nd in no_decay))
-                    # and param.requires_grad
-                ],
-                "lr": cfg.training.learning_rate,
-                "weight_decay": cfg.training.weight_decay,
-            },
-            {
-                "params": [
-                    param
-                    for name, param in model.named_parameters()
-                    if (not any(layer in name for layer in differential_layers))
-                    and (any(nd in name for nd in no_decay))
-                    # and param.requires_grad
-                ],
-                "lr": cfg.training.learning_rate,
-                "weight_decay": 0,
-            },
-            {
-                "params": [
-                    param
-                    for name, param in model.named_parameters()
-                    if (any(layer in name for layer in differential_layers))
-                    and (not any(nd in name for nd in no_decay))
-                    # and param.requires_grad
-                ],
-                "lr": cfg.training.differential_learning_rate,
-                "weight_decay": cfg.training.weight_decay,
-            },
-            {
-                "params": [
-                    param
-                    for name, param in model.named_parameters()
-                    if (any(layer in name for layer in differential_layers))
-                    and (any(nd in name for nd in no_decay))
-                    # and param.requires_grad
-                ],
-                "lr": cfg.training.differential_learning_rate,
-                "weight_decay": 0,
-            },
-        ],
-        lr=cfg.training.learning_rate,
-        weight_decay=cfg.training.weight_decay,
-    )
+    if cfg.environment.use_deepspeed and cfg.training.lora:
+        logger.info(
+            "Deepspeed /w Lora training do not support differential learning rate."
+        )
+        optimizer = Optimizers.get(cfg.training.optimizer)(
+            model.parameters(), lr=cfg.training.learning_rate
+        )
+    else:
+        no_decay = ["bias", "LayerNorm.weight"]
```

Comment:
how is `no_decay = ["bias", "LayerNorm.weight"]` handled with deepspeed?

if different, this needs documentation and warnings

Reply:
If I uncomment those `and param.requires_grad`
in this function it would work for deepspeed + lora as well.
Is there any special reason for commenting out `and param.requires_grad` ?

Reply:
Well, we don't apply weight decay to `["bias", "LayerNorm.weight"]`, that does make a difference.

Reply:
Deepspeed do not has special control on `["bias", "LayerNorm.weight"]`

I fix it as the following commit

[892f47c](https://github.com/h2oai/h2o-llmstudio/pull/288/commits/892f47cbfe6c56af96ffae737d5234fc825db332)